File: iec-xml/src/main/java/edu/kit/iti/formal/automation/plcopenxml/InterfaceBuilder.java
package edu.kit.iti.formal.automation.plcopenxml;

/*-
 * #%L
 * iec-xml
 * %%
 * Copyright (C) 2018 Alexander Weigl
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a clone of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */

import edu.kit.iti.formal.automation.IEC61131Facade;
import edu.kit.iti.formal.automation.scope.Scope;
import edu.kit.iti.formal.automation.st.ast.Literal;
import edu.kit.iti.formal.automation.st.ast.VariableBuilder;
import edu.kit.iti.formal.automation.st.ast.VariableDeclaration;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.RequiredArgsConstructor;
import org.jdom2.Element;

import java.util.function.Supplier;

/**
 * Construct from an &lt;interface&gt; and {@link edu.kit.iti.formal.automation.scope.Scope}
 *
 * @author Alexander Weigl
 * @version 1 (20.02.18)
 */
@Data
@RequiredArgsConstructor
@AllArgsConstructor
public class InterfaceBuilder implements Supplier<Scope> {
    private final Element interfaze;
    private Scope scope = new Scope();

    @Override
    public Scope get() {
        Scope d = new Scope();
        add(interfaze.getChild("inputVars"), d, VariableDeclaration.Companion.getINPUT());
        add(interfaze.getChild("outputVars"), d, VariableDeclaration.Companion.getOUTPUT());
        add(interfaze.getChild("localVars"), d, 0);
        //TODO Test for IN_OUT and others
        return d;
    }

    protected void add(Element vars, Scope d, int flags) {
        if (vars == null) {
            return;
        }

        VariableBuilder builder = d.builder();
        builder.push(flags);

        for (org.jdom2.Element e : vars.getChildren("variable")) {
            String name = e.getAttributeValue("name");
            String datatype = getDatatype(e.getChild("type"));
            Literal initValue = getInitialValue(e.getChild("initialValue"));
            builder.identifiers(name)
                    .setInitialization(initValue)
                    .setBaseType(datatype)
                    .create();
        }
    }

    protected String getDatatype(org.jdom2.Element e) {
        org.jdom2.Element derived = e.getChild("derived");
        if (derived != null) {
            return derived.getAttributeValue("name");
        } else {
            return e.getChildren().get(0).getName();
        }
    }


    private Literal getInitialValue(org.jdom2.Element initialValue) {
        if (initialValue == null)
            return null;

        org.jdom2.Element simpleValue = initialValue.getChild("simpleValue");
        if (simpleValue == null)
            return null;

        return (Literal) IEC61131Facade.INSTANCE.expr(simpleValue.getAttributeValue("value"));
    }
}

File: iec-xml/src/main/kotlin/edu/kit/iti/formal/automation/plcopenxml/InterfaceBuilder.kt
package edu.kit.iti.formal.automation.plcopenxml

import edu.kit.iti.formal.util.CodeWriter
import org.jdom2.Element

/**
 * Construct from an &lt;interface&gt; and [edu.kit.iti.formal.automation.scope.Scope]
 *
 * @author Alexander Weigl
 * @version 1 (20.02.18)
 */
object InterfaceBuilder : XMLTranslatorFind {
    override fun find(e: Element): Element? = when {
        e.name == "interface" -> e
        e.getChild("interface") != null -> e.getChild("interface")
        else -> null
    }

    override fun translate(e: Element, writer: CodeWriter) {
        val interfaze = find(e)
        translateVars(interfaze?.getChild("inputVars"), writer, "_INPUT")
        translateVars(interfaze?.getChild("outputVars"), writer, "_OUTPUT")
        translateVars(interfaze?.getChild("localVars"), writer)
        //TODO Test for IN_OUT and others
    }

    private fun translateVars(vars: Element?, writer: CodeWriter, suffix: String = "") {
        if (vars == null) return
        writer.nl().nl().append("VAR$suffix").block {
            for (e in vars.getChildren("variable")) {
                val name = e.getAttributeValue("name")
                val datatype = VariableDeclXML.getDatatype(e.getChild("type"))
                val initValue = VariableDeclXML.getInitialValue(e.getChild("initialValue"))
                nl()
                append("$name : $datatype")
                if (initValue != null)
                    append(":= $initValue")
                append(";")
            }
        }
        writer.nl().append("END_VAR")
    }

}


object VariableDeclXML {
    fun getDatatype(e: Element): String {
        val derived = e.getChild("derived")
        return if (derived != null) {
            derived.getAttributeValue("name")
        } else {
            e.children[0].name
        }
    }

    fun getInitialValue(initialValue: Element?): String? {
        if (initialValue == null) return null
        val simpleValue = initialValue.getChild("simpleValue") ?: return null
        return simpleValue.text
    }
}